### PR TITLE
WIP feat(actions): allow callable as action handlers

### LINKED ISSUE #10956: Developer Tools > Web Services callable description
Developer tools should describe method execution callables instead of echoing them

### LINKED ISSUE #11177: Input should be aware of Request
Router::route() takes in a request but Input has no clue what the request is. Either router needs to replace the request, or request needs to be injected in the input. IMO, we should just loose Input service and use the request.

### PR DESCRIPTION
**Pull request by hypeJunction**
Fixes #11177
Fixes #6760
Fixes #10956

### CODE REVIEW

File: docs/guides/actions.rst [hunk 1]
```diff
@@ -45,6 +73,9 @@ The location of the action files are assumed to be in the plugin folder  ``/acti
 		          'access' => 'admin',
 		          'filename' => __DIR__ . 'actions/blog/remove.php',
 		    ],
+            'blog/update' => [
+                  'callback' => \Elgg\Blog\Actions\UpdateAction::class,
+            ],
```
Comment (mrclay) [inline docs/guides/actions.rst]: Need to normalize tabs on this whole example.

File: engine/classes/Elgg/Action.php [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	 * @param string $key     The key of the value in the params array
+	 * @param mixed  $default The value to return if missing
+	 * @param bool   $filter  Sanitize input value
+	 *
+	 * @return mixed
+	 */
+	public function getParam($key, $default = null, $filter = true);
+
+	/**
+	 * Gets the "entity" key from the params if it holds an Elgg entity
```
Comment (mrclay) [inline engine/classes/Elgg/Action.php]: Does this get auto-provided from `get_input('guid')`?

File: engine/classes/Elgg/Actions/LoginAction.php [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+		$session->remove('last_forward_from');
+
+		$params = ['user' => $user, 'source' => $forward_source];
+		$forward_url = elgg_trigger_plugin_hook('login:forward', 'user', $params, $forward_url);
+
+		$output = [
+			'forward' => $forward_url,
+		];
+
+		if (elgg_is_xhr()) {
```
Comment (mrclay) [inline engine/classes/Elgg/Actions/LoginAction.php]: `$action->isXhr()` maybe? Obviously we don't need to bring in all Elgg API but this is a common one with actions.

File: engine/classes/Elgg/Action.php (hunk 2 above)
Comment (hypeJunction) [reply engine/classes/Elgg/Action.php]: Yes

File: engine/classes/Elgg/Actions/LoginAction.php (hunk 3 above)
Comment (hypeJunction) [reply engine/classes/Elgg/Actions/LoginAction.php]: Sure

**Comment by hypeJunction** [comment]
I don't want to push this for 3.0. Can go in later, no breaking changes

File: engine/classes/Elgg/ActionsService.php [hunk 4]
```diff
@@ -161,61 +180,80 @@ public function execute($action, $forwarder = "") {
...
+		$callback = false;
+
+		if (is_string($handler) && substr($handler, -4, 4) === '.php') {
+			if (!is_file($handler) || !is_readable($handler)) {
+				elgg_log("File $handler for action $action is not readable", 'ERROR');
+				return false;
+			}
+			$file = $handler;
+		} else {
+			$handler = $this->handlers->resolveCallable($handler);
```
Comment (hypeJunction) [inline engine/classes/Elgg/ActionsService.php]: This creates an instance, so maybe we should defer checking both files and callables until action is called

File: engine/classes/Elgg/ActionsService/Action.php [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+	 * {@inheritdoc}
+	 */
+	public function getParam($key, $default = null, $filter = true) {
+		return $this->input->get($key, $default, $filter);
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function getEntityParam() {
```
Comment (hypeJunction) [inline engine/classes/Elgg/ActionsService/Action.php]: Should accept key that holds entity guid.

File: engine/classes/Elgg/ActionsService/Action.php [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+			}
+		}
+
+		return null;
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function getUserParam() {
```
Comment (hypeJunction) [inline engine/classes/Elgg/ActionsService/Action.php]: Should accept key that holds user guid or username.

**Comment by hypeJunction** [comment]
Historically, actions could be functions, anyone know why that functionality has been removed?

File: engine/classes/Elgg/ActionsService.php [hunk 7]
```diff
@@ -161,61 +180,80 @@ public function execute($action, $forwarder = "") {
...
-
 		// set the maximum execution time for actions
 		$action_timeout = $this->config->action_time_limit;
 		if (isset($action_timeout)) {
 			set_time_limit($action_timeout);
 		}
 
-		$result = Includer::includeFile($file);
+		$callback = $this->actions[$action]['callback'];
+		if ($callback) {
```
Comment (mrclay) [inline engine/classes/Elgg/ActionsService.php]: If even filenames got wrapped in callback functions we wouldn't need this branch...

Comment (hypeJunction) [reply engine/classes/Elgg/ActionsService.php]: True

Comment (mrclay) [reply engine/classes/Elgg/ActionsService.php]: Naw. This results in far fewer objects created. It's fine.

**Comment by mrclay** [comment]
> Historically, actions could be functions, anyone know why that functionality has been removed?

I don't know. Maybe to nudge everyone towards one way of plugin dev...

**Comment by mrclay** [comment]
Some commits you can grab here: https://github.com/mrclay/Elgg-leaf/tree/hypeJunction-actions

**Comment by jdalsem** [comment]
@hypeJunction does this need to be in the beta release?

**Comment by hypeJunction** [comment]
No, there is nothing breaking here. Can go in later. The only thing lef to do for beta is update the remaining plugins' routes.

**Pull request closed without merging**